pkg/controllers/scheduling/scheduling_controller.go

```diff
@@ -0,0 +1,331 @@
+	if errors.IsNotFound(err) {
+		// no work if placementdecision is deleted
+		c.rateLimiter.Forget(queueKey)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// no work if placementdecision is deleting
+	if !placementDecision.DeletionTimestamp.IsZero() {
+		c.rateLimiter.Forget(queueKey)
+		return nil
+	}
+
+	return c.schedule(ctx, syncCtx, placementDecision)
+}
+
+// schedule makes decisions for the unscheduled decisions in the status of PlacementDecision.
+func (c *decisionSchedulingController) schedule(ctx context.Context, syncCtx factory.SyncContext, placementDecision *clusterapiv1alpha1.PlacementDecision) error {
+	selectedClusters := sets.NewString()
+	for _, decision := range placementDecision.Status.Decisions {
+		if len(decision.ClusterName) == 0 {
+			continue
+		}
+
+		selectedClusters.Insert(decision.ClusterName)
+	}
+
+	// no work if all decisions have already been scheduled
+	queueKey := fmt.Sprintf("%s/%s", placementDecision.Namespace, placementDecision.Name)
+	numOfDecisions := len(placementDecision.Status.Decisions)
+	if numOfDecisions == selectedClusters.Len() {
+		c.rateLimiter.Forget(queueKey)
+		return nil
+	}
+
+	// get the owner placement
+	placementName := placementDecision.Labels[placementLabel]
+	if len(placementName) == 0 {
+		return fmt.Errorf("no placement label on placementdecision %q", queueKey)
+	}
+	placement, err := c.placementLister.Placements(placementDecision.Namespace).Get(placementName)
+	if err != nil {
+		return err
+	}
+
+	// get all feasible clusters for the placement
+	clusters, err := getFeasibleClusters(placement, c.clusterLister, c.clusterSetLister, c.clusterSetBindingLister)
+	if err != nil {
+		return err
+	}
+
+	// make decisions for the unscheduled decisions
+	newPlacementDecision := placementDecision.DeepCopy()
+	for index, decision := range newPlacementDecision.Status.Decisions {
+		if len(decision.ClusterName) != 0 {
+			continue
+		}
+
+		decision := scheduleOne(clusters, selectedClusters)
+		// stop if no more cluster is available
+		if decision == nil {
+			break
+		}
+
+		newPlacementDecision.Status.Decisions[index] = *decision
+		selectedClusters.Insert(decision.ClusterName)
+	}
+
+	// update the placementdecision if necessary
+	if reflect.DeepEqual(placementDecision.Status.Decisions, newPlacementDecision.Status.Decisions) {
+		c.requeueOrForget(queueKey, syncCtx.Queue(), numOfDecisions, selectedClusters.Len())
+		return nil
+	}
+
+	if _, err := c.clusterClient.ClusterV1alpha1().PlacementDecisions(newPlacementDecision.Namespace).
+		UpdateStatus(ctx, newPlacementDecision, metav1.UpdateOptions{}); err != nil {
+		return err
+	}
+
+	// update the status of the placement if necessary
+	if err := updatePlacementStatus(ctx, placement, numOfDecisions, selectedClusters.Len(), c.clusterClient); err != nil {
+		return err
+	}
+
+	c.requeueOrForget(queueKey, syncCtx.Queue(), numOfDecisions, selectedClusters.Len())
```

Comment:
why we  need this again?

Reply:
This function should be invoked before schedule() ends without error. To be specific,
- PlacementDecision is not changed; (the first call)
- PlacementDecision is changed; (the second call)

Reply:
Code updated as we discussed offline